### PR TITLE
Better errors when JtR version installed from package isn't adequate

### PR DESCRIPTION
**Pull request by pbarry-r7**
John the Ripper installed from native packages, like ubuntu's `john` package, might output the version in a slightly different format than when you build JtR from source.  This patch adds an additional attempt to pull the version number from JtR's like these...

I noticed this recently while following test/usage examples in #11695 .  This patch is intended to be unobtrusive and works against Ubuntu 16.04 with JtR installed via the standard apt-get utility (I did not test other platforms), while not impacting the check for JUMBO JtR.

## Existing behavior when `john` is installed in Ubuntu 16.04, but not the JUMBO JtR

```
msf5 > creds add user:des2_password hash:rEK1ecacw.7.c jtr:des
msf5 > creds add user:des_password hash:rEK1ecacw.7.c jtr:des
msf5 > creds add user:des_55 hash:rDpJV6xlcXxRM jtr:des
msf5 > creds add user:des_pot_55 hash:fakeV6xlcXxRM jtr:des
msf5 > creds add user:des_passphrase hash:qiyh4XPJGsOZ2MEAyLkfWqeQ jtr:des
msf5 > echo "fakeV6xlcXxRM:55" >> /home/vagrant/.msf4/john.pot
[*] exec: echo "fakeV6xlcXxRM:55" >> /home/vagrant/.msf4/john.pot

msf5 > echo "test" > /tmp/wordlist
[*] exec: echo "test" > /tmp/wordlist

msf5 > echo "password" >> /tmp/wordlist
[*] exec: echo "password" >> /tmp/wordlist

msf5 > setg CUSTOM_WORDLIST /tmp/wordlist
CUSTOM_WORDLIST => /tmp/wordlist
msf5 > setg ShowCommand true
ShowCommand => true
msf5 > setg USE_DEFAULT_WORDLIST false
USE_DEFAULT_WORDLIST => false
msf5 > setg DeleteTempFiles false
DeleteTempFiles => false
msf5 > setg USE_CREDS false
USE_CREDS => false
msf5 > setg USE_DB_INFO false
USE_DB_INFO => false
msf5 > setg USE_HOSTNAMES false
USE_HOSTNAMES => false
msf5 > setg USE_ROOT_WORDS false
USE_ROOT_WORDS => false
msf5 > setg ITERATION_TIMEOUT 60
ITERATION_TIMEOUT => 60
msf5 > use auxiliary/analyze/crack_aix
msf5 auxiliary(analyze/crack_aix) > run

[-] Auxiliary failed: NoMethodError undefined method `include?' for nil:NilClass
[-] Call stack:
[-]   /vagrant/modules/auxiliary/analyze/crack_aix.rb:124:in `run'
[*] Auxiliary module execution completed
```

## With the patch applied

```
msf5 > creds add user:des2_password hash:rEK1ecacw.7.c jtr:des
msf5 > creds add user:des_password hash:rEK1ecacw.7.c jtr:des
msf5 > creds add user:des_55 hash:rDpJV6xlcXxRM jtr:des
msf5 > creds add user:des_pot_55 hash:fakeV6xlcXxRM jtr:des
msf5 > creds add user:des_passphrase hash:qiyh4XPJGsOZ2MEAyLkfWqeQ jtr:des
msf5 > echo "fakeV6xlcXxRM:55" >> /home/vagrant/.msf4/john.pot
[*] exec: echo "fakeV6xlcXxRM:55" >> /home/vagrant/.msf4/john.pot

msf5 > echo "test" > /tmp/wordlist
[*] exec: echo "test" > /tmp/wordlist

msf5 > echo "password" >> /tmp/wordlist
[*] exec: echo "password" >> /tmp/wordlist

msf5 > setg CUSTOM_WORDLIST /tmp/wordlist
CUSTOM_WORDLIST => /tmp/wordlist
msf5 > setg ShowCommand true
ShowCommand => true
msf5 > setg USE_DEFAULT_WORDLIST false
USE_DEFAULT_WORDLIST => false
msf5 > setg DeleteTempFiles false
DeleteTempFiles => false
msf5 > setg USE_CREDS false
USE_CREDS => false
msf5 > setg USE_DB_INFO false
USE_DB_INFO => false
msf5 > setg USE_HOSTNAMES false
USE_HOSTNAMES => false
msf5 > setg USE_ROOT_WORDS false
USE_ROOT_WORDS => false
msf5 > setg ITERATION_TIMEOUT 60
ITERATION_TIMEOUT => 60
msf5 > use auxiliary/analyze/crack_aix
msf5 auxiliary(analyze/crack_aix) > run

[-] Auxiliary aborted due to failure: bad-config: John the Ripper JUMBO patch version required.  See https://github.com/magnumripper/JohnTheRipper
[*] Auxiliary module execution completed
```

## Verification

List the steps needed to make sure this thing works

- [ ] Begin with a system that has a non-JUMBO version of JtR installed
- [ ] Start `msfconsole`
- [ ] Follow an example in the module docs for testing one of the crack password modules (#11695)
- [ ] **Verify** you get an appropriate, descriptive error that you aren't using a JUMBO version of JtR
- [ ] Install a [JUMBO version](https://github.com/magnumripper/JohnTheRipper/blob/bleeding-jumbo/doc/INSTALL-UBUNTU) of JtR
- [ ] Rerun your same example (maybe set the CRACKER_PATH module option if you need to point to your new JUMBO `john` executable)
- [ ] **Verify** you do NOT get any error about missing JUMBO JtR and that password cracking commences

### CODE REVIEW

**Comment by h00die** [comment]
what was the version string out of curiousity

**Comment by bcoles** [comment]
It appears `John the Ripper password cracker, version 1.8.0.2-bleeding-jumbo_omp
[64-bit AVX-autoconf]` is also a [valid format](https://john-dev.openwall.narkive.com/KflCd9T9/autoconf-and-list-build-info).

Also, `(.+)` regex should perhaps be less greedy `(.+?)`.

**Comment by pbarry-r7** [comment]
Hey @h00die, apologies for the late reply.  In my situation, the john package installed via apt-get looks like this:

```
$ john 
John the Ripper password cracker, version 1.8.0
Copyright (c) 1996-2013 by Solar Designer
Homepage: http://www.openwall.com/john/

Usage: john [OPTIONS] [PASSWORD-FILES]
--single                   "single crack" mode
--wordlist=FILE --stdin    wordlist mode, read words from FILE or stdin
--rules                    enable word mangling rules for wordlist mode
--incremental[=MODE]       "incremental" mode [using section MODE]
--external=MODE            external mode or word filter
--stdout[=LENGTH]          just output candidate passwords [cut at LENGTH]
--restore[=NAME]           restore an interrupted session [called NAME]
--session=NAME             give a new session the NAME
--status[=NAME]            print status of a session [called NAME]
--make-charset=FILE        make a charset, FILE will be overwritten
--show                     show cracked passwords
--test[=TIME]              run tests and benchmarks for TIME seconds each
--users=[-]LOGIN|UID[,..]  [do not] load this (these) user(s) only
--groups=[-]GID[,..]       load users [not] of this (these) group(s) only
--shells=[-]SHELL[,..]     load users with[out] this (these) shell(s) only
--salts=[-]N               load salts with[out] at least N passwords only
--save-memory=LEVEL        enable memory saving, at LEVEL 1..3
--node=MIN[-MAX]/TOTAL     this node's number range out of TOTAL count
--fork=N                   fork N processes
--format=NAME              force hash type NAME: descrypt/bsdicrypt/md5crypt/
                           bcrypt/LM/AFS/tripcode/dummy/crypt
```

**Comment by h00die** [comment]
I'm trying to find a good regex to match all the test cases.  https://rubular.com/r/XY0pYc5cWDf0D3
That matches on what @bcoles found as well, but still missing on yours @pbarry-r7 .  Maybe a final match on [ or end of line?

**Comment by pbarry-r7** [comment]
Ah, I have not seen rubular, neat!  Appreciate y'all's thoughts on this, @h00die and @bcoles.   I believe this regex works for the 4 test cases: https://rubular.com/r/vf3W1dJ85qKrY7.  I tested it in my code against those reported JtR versions, and it does appear to be returning the version for all 4 cases (and then correctly balking on my non-JUMBO version string, yay!).

**Comment by pbarry-r7** [comment]
Meh, looks like it leaves a trailing character on the match (space or carriage return, depending on match).  :/

File: lib/metasploit/framework/password_crackers/cracker.rb [hunk 1]
```diff
@@ -320,8 +320,9 @@ def cracker_version
               if cracker == 'john'
                 # John the Ripper 1.8.0.13-jumbo-1-bleeding-973a245b96 2018-12-17 20:12:51 +0100 OMP [linux-gnu 64-bit x86_64 AVX2 AC]
                 # John the Ripper 1.9.0-jumbo-1 OMP [linux-gnu 64-bit x86_64 AVX2 AC]
-                #return $1 if line =~ /John the Ripper ([\.\w-]+) 20\d{2}-\d{2}-\d{2}/
-                return $1 if line =~ /John the Ripper (.+) \[/
+                # John the Ripper password cracker, version 1.8.0.2-bleeding-jumbo_omp [64-bit AVX-autoconf]
+                # John the Ripper password cracker, version 1.8.0
+                return $1 if line =~ /(?: password cracker, version)? ([^\[]+)?/
```
Comment (pbarry-r7) [inline lib/metasploit/framework/password_crackers/cracker.rb]: ....and Jeffrey pointed out that I also missed the "John the Ripper" match here when I updated the code...  I should have more coffee in the mornings before delving into regex...

**Comment by jmartin-r7** [comment]
@h00die is there are reason we need anything more that the first token of the version string?  I do not see anywhere we do anything with the `OMP` portion when it is included.

If there is a reason we need more details on the capabilities of a cracker in the future I would be valuable to give that its own method I think.

**Comment by pbarry-r7** [comment]
Latest commit here re-added the ''JtR' match I had unintentionally dropped this morning and does strip off any trailing spaces/newlines (per testing).

**Comment by h00die** [comment]
@jmartin-r7 nope.  There were two purposes behind including this at all:

1. since its an external uncontrolled (gem) item, when someone inevitably has an issue, we want to know what version they are running to help debug
2. ensure 'jumbo' is part of the version string since the non-jumbo doesn't have MANY of the hash types we need/want.  In theory it may be possible to run **some** of the cracker modules w/o the jumbo patch, but I assume it has optimizations to those non-jumbo hashes as well, so it's just bes tto have jumbo.  (https://github.com/rapid7/metasploit-framework/blob/eb0730510421700926bff9a6a40750442a4f1ca5/modules/auxiliary/analyze/crack_linux.rb#L142)

Part of this originated from here: https://github.com/rapid7/metasploit-framework/issues/11294#issuecomment-473149082

**Comment by jmartin-r7** [comment]
## Release Notes
Improved detection and reporting related to the JtR version available.

**Pull request merged**